File: systems/crawler/backend/src/main/java/com/aacv/system/catalog/domain/CatalogQuery.java
package com.aacv.system.catalog.domain;

public record CatalogQuery(
        String title,
        String author,
        String organization,
        Integer publicationYear,
        String achievementType,
        String sourceCode,
        String venue,
        String topic,
        int page,
        int size,
        Long authorId,
        Long organizationId,
        Long venueId,
        Long topicId) {

    public CatalogQuery(String title, String author, String organization, Integer publicationYear,
            String achievementType, String sourceCode, String venue, String topic, int page, int size) {
        this(title, author, organization, publicationYear, achievementType, sourceCode, venue, topic,
                page, size, null, null, null, null);
    }

    // 目录和导出共用筛选 SQL；目录的单年条件等价于闭合年份区间。
    public Integer publicationYearFrom() { return publicationYear; }
    public Integer publicationYearTo() { return publicationYear; }
    public String sourceType() { return null; }

    public CatalogQuery {
        for (Long id : new Long[] {authorId, organizationId, venueId, topicId}) {
            if (id != null && id < 1) throw new IllegalArgumentException("检索实体ID无效");
        }
        if (publicationYear != null && (publicationYear < 1000 || publicationYear > 9999)) {
            throw new IllegalArgumentException("发表年份无效");
        }
        if (page < 0 || size < 1 || size > 100) {
            throw new IllegalArgumentException("分页参数无效");
        }
        title = normalize(title);
        author = normalize(author);
        organization = normalize(organization);
        achievementType = normalize(achievementType);
        sourceCode = normalize(sourceCode);
        venue = normalize(venue);
        topic = normalize(topic);
    }

    private static String normalize(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        String normalized = value.trim();
        if (normalized.length() > 200) {
            throw new IllegalArgumentException("检索条件长度超出限制");
        }
        return normalized;
    }
}

File: systems/crawler/frontend/src/utils/catalog-query.ts
import type { LocationQuery, LocationQueryRaw } from 'vue-router'
import type { AchievementQuery } from '@/services/business'

const textKeys = ['title', 'author', 'organization', 'venue', 'topic'] as const
const idKeys = ['authorId', 'organizationId', 'venueId', 'topicId'] as const

/** URL 只接受目录已知条件，避免详情返回时携带无关参数。 */
export function readCatalogQuery(query: LocationQuery): AchievementQuery {
  const result: AchievementQuery = { page: integer(query.page, 0, 1000000) ?? 0, size: integer(query.size, 1, 100) ?? 20 }
  for (const key of textKeys) result[key] = typeof query[key] === 'string' ? query[key].trim().slice(0, 200) : ''
  for (const key of idKeys) result[key] = integer(query[key], 1, Number.MAX_SAFE_INTEGER)
  result.publicationYear = integer(query.publicationYear, 1000, 9999)
  return result
}

export function catalogRouteQuery(query: AchievementQuery): LocationQueryRaw {
  return Object.fromEntries(Object.entries(query)
    .filter(([key]) => [...textKeys, ...idKeys, 'publicationYear', 'page', 'size'].includes(key))
    .filter(([key, value]) => value !== undefined && value !== '' && !(key === 'page' && value === 0) && !(key === 'size' && value === 20))
    .map(([key, value]) => [key, String(value)]))
}

function integer(value: unknown, min: number, max: number): number | undefined {
  if (typeof value !== 'string' || !/^\d+$/.test(value)) return undefined
  const number = Number(value)
  return Number.isSafeInteger(number) && number >= min && number <= max ? number : undefined
}
